<a href="https://colab.research.google.com/github/raushan9jnv/Global-Terrorism-Analysis/blob/main/FINAL_Global_Terrorism_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

**Accessing the data**
> Mounting the data from Google drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
file_path = '/content/drive/MyDrive/Data Science/Python/Global Terrorism analysis/Global Terrorism Data.csv'

In [ ]:
df = pd.read_csv(file_path,encoding="ISO-8859-1",engine='python')

**Reading the dataset**
> First we try to read the data from every aspect like Total no of columns, Total no of rows, DataType in each column, Null values, Duplicate values and WHAT DATA TALKING ABOUT. 
 this small analysis will give us a generalized idea about with what types of data we dealing.
 > let's do it :)

*Access full Dataset*

In [ ]:
df

**First 5 rows of our DataFrame**
> our data is comparably big i.e. 181691 rows, so it will be a good pratice to see small set of rows

In [ ]:
df.head()

**Dimension of dataset**

In [ ]:
df.shape

(181691, 135)

From above we saw their total 181691 ROWS and 135 COLUMNS. 
> Now let's see which type of information stored in columns.

**Columns Values**

In [ ]:
df.columns  

Index(['eventid', 'iyear', 'imonth', 'iday', 'approxdate', 'extended',
       'resolution', 'country', 'country_txt', 'region',
       ...
       'addnotes', 'scite1', 'scite2', 'scite3', 'dbsource', 'INT_LOG',
       'INT_IDEO', 'INT_MISC', 'INT_ANY', 'related'],
      dtype='object', length=135)

In [ ]:
#Print full column Names
df.columns.tolist() 

**DataTypes of each column**

In [ ]:
df.dtypes

In [ ]:
df.dtypes.tolist()

**Full summary of DataSet**

In [ ]:
df.describe()

In [ ]:
df.info()

**Null Values in DataFrame**

In [ ]:
df.isnull()


In [ ]:
df.isnull().sum()

**Total Number of Null Values**

In [ ]:
df.isnull().sum().sum()

13853997

> it quiet intresting,isn't it. we handling such large number of data  and from their 13853997 values are unknow.
> Further we drop these all NAN values for our ease and precisely analysis.

**Years in our DataFrame**

In [ ]:
df.nunique()

In [ ]:
df['iyear'].unique()

In [ ]:
df['iyear'].nunique()

47

> This cell clearly conclude that we have to work on total 47 Years from taking 1970 to 2017. this all was years of attack on different region.

**WHAT WE OBSERVE TILL NOW**

> 1. Their total 181691 rows and 135 columns.  ==> large data information
>2.  We have to deal with all three general types of datatype ==> int64, float64 and object
> 3. Most of columns are irrelevent repect to our analysis  ==>  we drop them further
> 4.  So many NAN and Unknow values i.e. 13853997 ==> we also drop them
> 5. Total 47 years  data information ==> 1970 to 2017

#PRE-PROCESSINNG OF DATA

**Dropping Columns**

> keep the columns which are relevate to us and drop rest.

In [ ]:
usecolumns = {1,5,8,10,11,12,13,14,25,26,27,29,35,58,69,71,82,98,100,101,103,104,106}
df = pd.read_csv(file_path,encoding="ISO-8859-1",engine='python',usecols=usecolumns)
print('we successfully reduce column from 135 to 23')

we successfully reduce column from 135 to 23


In [ ]:
df

iyear  extended         country_txt                   region_txt  \
0        1970         0  Dominican Republic  Central America & Caribbean   
1        1970         0              Mexico                North America   
2        1970         0         Philippines               Southeast Asia   
3        1970         0              Greece               Western Europe   
4        1970         0               Japan                    East Asia   
...       ...       ...                 ...                          ...   
181686   2017         0             Somalia           Sub-Saharan Africa   
181687   2017         0               Syria   Middle East & North Africa   
181688   2017         0         Philippines               Southeast Asia   
181689   2017         0               India                   South Asia   
181690   2017         0         Philippines               Southeast Asia   

              provstate           city   latitude   longitude  multiple  \
0                   NaN  Santo Domingo  18.456792  -69.951164       0.0   
1               Federal    Mexico city  19.371887  -99.086624       0.0   
2                Tarlac        Unknown  15.478598  120.599741       0.0   
3                Attica         Athens  37.997490   23.762728       0.0   
4               Fukouka        Fukouka  33.580412  130.396361       0.0   
...                 ...            ...        ...         ...       ...   
181686  Middle Shebelle  Ceelka Geelow   2.359673   45.385034       0.0   
181687         Lattakia         Jableh  35.407278   35.942679       0.0   
181688      Maguindanao       Kubentog   6.900742  124.437908       0.0   
181689          Manipur         Imphal  24.798346   93.940430       0.0   
181690      Maguindanao  Cotabato City   7.209594  124.241966       0.0   

        success  suicide                 attacktype1_txt  \
0             1        0                   Assassination   
1             1        0     Hostage Taking (Kidnapping)   
2             1        0                   Assassination   
3             1        0               Bombing/Explosion   
4             1        0  Facility/Infrastructure Attack   
...         ...      ...                             ...   
181686        1        0                   Armed Assault   
181687        1        0               Bombing/Explosion   
181688        1        0  Facility/Infrastructure Attack   
181689        0        0               Bombing/Explosion   
181690        0        0               Bombing/Explosion   

                      targtype1_txt  \
0       Private Citizens & Property   
1           Government (Diplomatic)   
2               Journalists & Media   
3           Government (Diplomatic)   
4           Government (Diplomatic)   
...                             ...   
181686                     Military   
181687                     Military   
181688  Private Citizens & Property   
181689         Government (General)   
181690                      Unknown   

                                             gname  nperps  claimed  \
0                                           MANO-D     NaN      NaN   
1               23rd of September Communist League     7.0      NaN   
2                                          Unknown     NaN      NaN   
3                                          Unknown     NaN      NaN   
4                                          Unknown     NaN      NaN   
...                                            ...     ...      ...   
181686                                  Al-Shabaab   -99.0      1.0   
181687                           Muslim extremists   -99.0      0.0   
181688  Bangsamoro Islamic Freedom Movement (BIFM)   -99.0      0.0   
181689                                     Unknown   -99.0      0.0   
181690                                     Unknown   -99.0      0.0   

       weaptype1_txt  nkill  nkillter  nwound  nwoundte  property  \
0            Unknown    1.0       NaN     0.0       NaN         0   
1            

In [ ]:
df.head()

iyear  extended         country_txt                   region_txt provstate  \
0   1970         0  Dominican Republic  Central America & Caribbean       NaN   
1   1970         0              Mexico                North America   Federal   
2   1970         0         Philippines               Southeast Asia    Tarlac   
3   1970         0              Greece               Western Europe    Attica   
4   1970         0               Japan                    East Asia   Fukouka   

            city   latitude   longitude  multiple  success  suicide  \
0  Santo Domingo  18.456792  -69.951164       0.0        1        0   
1    Mexico city  19.371887  -99.086624       0.0        1        0   
2        Unknown  15.478598  120.599741       0.0        1        0   
3         Athens  37.997490   23.762728       0.0        1        0   
4        Fukouka  33.580412  130.396361       0.0        1        0   

                  attacktype1_txt                targtype1_txt  \
0                   Assassination  Private Citizens & Property   
1     Hostage Taking (Kidnapping)      Government (Diplomatic)   
2                   Assassination          Journalists & Media   
3               Bombing/Explosion      Government (Diplomatic)   
4  Facility/Infrastructure Attack      Government (Diplomatic)   

                                gname  nperps  claimed weaptype1_txt  nkill  \
0                              MANO-D     NaN      NaN       Unknown    1.0   
1  23rd of September Communist League     7.0      NaN       Unknown    0.0   
2                             Unknown     NaN      NaN       Unknown    1.0   
3                             Unknown     NaN      NaN    Explosives    NaN   
4                             Unknown     NaN      NaN    Incendiary    NaN   

   nkillter  nwound  nwoundte  property propextent_txt  
0       NaN     0.0       NaN         0            NaN  
1       NaN     0.0       NaN         0            NaN  
2       NaN     0.0       NaN         0            NaN  
3       NaN     NaN       NaN         1            NaN  
4       NaN     NaN       NaN         1            NaN

**Dimension of Updated DataFrame**

In [ ]:
df.shape

(181691, 23)

**Columns we keeped**

In [ ]:
df.columns

Index(['iyear', 'extended', 'country_txt', 'region_txt', 'provstate', 'city',
       'latitude', 'longitude', 'multiple', 'success', 'suicide',
       'attacktype1_txt', 'targtype1_txt', 'gname', 'nperps', 'claimed',
       'weaptype1_txt', 'nkill', 'nkillter', 'nwound', 'nwoundte', 'property',
       'propextent_txt'],
      dtype='object')

**Remove unknows value in Co-ordinate**

In [ ]:
df=df[pd.notnull(df.latitude)]
df=df[pd.notnull(df.longitude)]
print('unknows value in coordinates are removed successfully')

unknows value in coordinates are removed successfully


In [ ]:
df.dtypes

iyear                int64
extended             int64
country_txt         object
region_txt          object
provstate           object
city                object
latitude           float64
longitude          float64
multiple           float64
success              int64
suicide              int64
attacktype1_txt     object
targtype1_txt       object
gname               object
nperps             float64
claimed            float64
weaptype1_txt       object
nkill              float64
nkillter           float64
nwound             float64
nwoundte           float64
property             int64
propextent_txt      object
dtype: object

**Unknows in numric columns**

In [ ]:
exclude_cols= ['iyear','latitude','longitude']
float_cols =[c for c in df.select_dtypes(include=[float]).columns.tolist() if c not in exclude_cols]

> We dont want any NANs as they convert the column to dtype = float

**Convert unknows into Negative**

In [ ]:
df[float_cols] = df[float_cols].fillna(0).astype(int)
df[float_cols] = df[float_cols].mask(df[float_cols]<0,0)

**Unknows in string columns**

In [ ]:
str_cols = df.select_dtypes(exclude =[float]).columns.tolist()
df[str_cols] =df[str_cols].fillna('unknow')
# Many fields are dots which mean unknown values
df[str_cols] = df[str_cols].replace(r'^\,*$','unknown',regex= True)

**Number of duplicates**

In [ ]:
df.duplicated().sum()

21734

> 1. After manuplating data, 21734 duplicate values found.
> 2. we should drop the duplicates values.

**Removal of duplicates values**

In [ ]:
df.drop_duplicates(keep=False,inplace=True)

**is there still Null values? let's check!**

In [ ]:
df.isnull().sum().sum()

0

> There no null values found. Now we are ready to get useful information from Dataset.

# Gathering Information

**Let's try to find out**
> 1. Which year terrorist attacked most?
> 2. Which Region and affected most by attack?
> 3. Which city affected most by attack?
> 4. What type of method terrorist used to attack?
> 5. Most  type of Target attacked by terrorist?
> 6. Which gang found regularly active?
> 7. Is there any suicide happened? and 
> 8. Number of successful attack planned by terrorist?
> 9. Is there any Unsuccessful attack by terrorist?

In [ ]:
df

In [ ]:
df.info()

In [ ]:
df.describe()

**Unique values in each column of dataset**

In [ ]:
df.nunique()

**1. Which year terrorist attacked most?**

In [ ]:
df['iyear'].value_counts()

In [ ]:
print(f"Maximum number of Attacked happened in Year 2014 and it was total {df['iyear'].value_counts().max()}.")
print(f"The least number of Attacked happened in Year 1972 and it was total {df['iyear'].value_counts().min()}.")

**2. Which Region affected most by attack?**

In [ ]:
df['region_txt'].value_counts()

In [ ]:
print(f"Maximum number of Attacked happened in Region SOUTH ASIA and it was total {df['region_txt'].value_counts().max()}.")
print(f"The least number of Attacked happened in Region Australasia & Oceania and it was total {df['region_txt'].value_counts().min()}.")

**3. Which city affected most by attack?**

In [ ]:
df["city"].value_counts()

In [ ]:
print(f"Most affacted city was Unknown and it was attacked {df['city'].value_counts().max()} times.")
print(f"The least affected city was Sakadra and it was attacked {df['city'].value_counts().min()} times.")

**4. What type of method terrorist used to attack most?**

In [ ]:
df["attacktype1_txt"].value_counts()

In [ ]:
print(f"The most used attacked method was Bombing/Explosion and it used {df['attacktype1_txt'].value_counts().max()} times.")
print(f"The least attacked method was Hijacking and it was used {df['attacktype1_txt'].value_counts().min()} times.")

**5. Most type of Targetted attacked by terrorist?**

In [ ]:
df["targtype1_txt"].value_counts()

In [ ]:
print(f"The most Targetted type was Private Citizens & Property and it attacked {df['targtype1_txt'].value_counts().max()} times.")
print(f"The least attacked method was Other and it was attacked {df['targtype1_txt'].value_counts().min()} times.")

The most Targetted type was Private Citizens & Property and it attacked 34521 times.
The least attacked method was Other and it was attacked 122 times.


**6. Which gang found regularly active?**

In [ ]:
df["gname"].value_counts()

Unknown                                              65968
Taliban                                               7130
Islamic State of Iraq and the Levant (ISIL)           4636
Al-Shabaab                                            2854
Shining Path (SL)                                     2702
                                                     ...  
Gholam Yahya Akbar                                       1
Gaddafi loyalists                                        1
Youth Movement for the Total Liberation of Azawad        1
Togolese Peoples Rally                                   1
Revolutionary Action Groups                              1
Name: gname, Length: 3398, dtype: int64

In [ ]:
print(f"The regularly active gang was Unknown and it attacked {df['gname'].value_counts().max()} times.")
print(f"The second most active gang was TALIBAN and it attacked 7130 times")
print(f"The least active gang was Revolutionary Action Groups and it attacked {df['gname'].value_counts().min()} times.")

The regularly active gang was Unknown and it attacked 65968 times.
The second most active gang was TALIBAN and it attacked 7130 times
The least active gang was Revolutionary Action Groups and it attacked 1 times.


**7. Is there any suicide happened?**

In [ ]:
df['suicide'].value_counts()

0    139753
1      6388
Name: suicide, dtype: int64

In [ ]:
print("Yes. The total number of suicide happened in these 47 years was 6388.")

Yes. The total number of suicide happened in these 47 years was 6388.


**8. Number of successful attack planned by terrorist?**

In [ ]:
df["success"].value_counts()

1    129477
0     16664
Name: success, dtype: int64

In [ ]:
print("The total successful attacked in 47 years was 129477.")

The total successful attacked in 47 years was 129477.


**9. Number of successful attack by terrorist?**

In [ ]:
print("YES. The total UNsuccessful attacked in 47 years was 16664.")

YES. The total UNsuccessful attacked in 47 years was 16664.


>Now we have Answers of all questions 

In [ ]:
df.corr()

In [ ]:
df.cov()

#visualization of data

In [ ]:
fig,axes = plt.subplots(1,1,figsize=(16,14))
sns.heatmap(df.corr(),annot=True)
plt.show()

In [ ]:
fig,axes = plt.subplots(1,1,figsize=(16,14))
sns.heatmap(df.cov(), annot=True)
plt.show()